// rrweb/es/rrweb/packages/rrweb/src/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./record/index.js');
const index$1 = require('./replay/index.js');
const utils = require('./utils.js');
const types = require('./types.js');

var addCustomEvent = index.default.addCustomEvent;
var freezePage = index.default.freezePage;

exports.record = index.default;
exports.Replayer = index$1.Replayer;
Object.defineProperty(exports, 'mirror', {
	enumerable: true,
	get: () => utils._mirror
});
exports.utils = utils;
Object.defineProperty(exports, 'EventType', {
	enumerable: true,
	get: () => types.EventType
});
Object.defineProperty(exports, 'IncrementalSource', {
	enumerable: true,
	get: () => types.IncrementalSource
});
Object.defineProperty(exports, 'MouseInteractions', {
	enumerable: true,
	get: () => types.MouseInteractions
});
Object.defineProperty(exports, 'ReplayerEvents', {
	enumerable: true,
	get: () => types.ReplayerEvents
});
exports.addCustomEvent = addCustomEvent;
exports.freezePage = freezePage;
